src/generators/operationGenerator.ts

```diff
@@ -339,24 +435,39 @@ function getOperationParameterSignatures(
         usedModels.forEach(model => importedModels.add(model));
       }
 
-      return {
+      const newParameter = {
         name: param.name,
         description: param.description,
         type: typeName,
         hasQuestionToken: !param.required
       };
-    });
 
-    // add optional parameter
-    const optionalParameter = getOptionsParameter(
+      // Make sure required parameters are added before optional
+      return param.required ? [newParameter, ...acc] : [...acc, newParameter];
+    }, []);
+
+    trackParameterGroups(
       operation,
       parameters,
       importedModels,
-      {
-        mediaType: hasMultipleOverloads ? requestMediaType : undefined
-      }
+      parameterDeclarations
     );
-    parameterDeclarations.push(optionalParameter);
+
+    // Only use our optional parameter grouping if the parameter grouping extension
+    // is not used in the swagger, if so, honor the extension grouping.
```

Comment:
I'm a little confused by this comment. Are we saying that if parameter grouping is enabled for any parameter in an operation, we won't put any non-grouped optional parameters in our options bag? I'm wondering if we would still want to put optional parameters in the `options` object if they aren't part of a parameter group.

Let's take this Swagger example:
```js
"/some/{pathParam1}/{pathParam2}": {
  "operationId": "Update",
  "post": {
    "parameters": [
    {
        "name": "headerParam",
        "in": "header",
        "type": "string",
        "required": false,
        "x-ms-parameter-grouping": {
          "name": "custom-parameter-group"
        }
    },
    {
        "name": "pathParam",
        "in": "path",
        "type": "string",
        "required": true,
        "x-ms-parameter-grouping": {
          "name": "custom-parameter-group"
        }
    },
    {
        "name": "headerParam2",
        "in": "path",
        "type": "string",
        "required": false,
    }]
  }
}
```
In this case, would we get a function like:
```ts
interface CustomParameterGroup {
  headerParam?: string;
  pathParam: string;
}
interface UpdateOptions extends OperationOptions {
  headerParam2?: string
}
client.update({headerParam, pathParam}: CustomParameterGroup, options);
```

Reply:
Thanks for the comment @chradek, this is very reasonable. As we discussed offline we'll proceed with:

* Keep doing our optional parameter grouping, even when parameter grouping extension is used
* If a group is optional, we'll put it in the Options bag